packages/block-library/src/block/edit.js

```diff
@@ -132,11 +133,36 @@ function getOverridesFromBlocks( blocks, defaultValues ) {
 	return Object.keys( overrides ).length > 0 ? overrides : undefined;
 }
 
+function setBlockEditMode( setEditMode, blocks ) {
+	blocks.forEach( ( block ) => {
+		const editMode = isPartiallySynced( block )
+			? 'contentOnly'
+			: 'disabled';
+		setEditMode( block.clientId, editMode );
+		setBlockEditMode( setEditMode, block.innerBlocks );
+	} );
+}
+
+function editSourcePattern( editInPatternOnlyMode, patternId ) {
+	const currentArgs = getQueryArgs( window.location.href );
+	const currentUrlWithoutArgs = removeQueryArgs(
+		window.location.href,
+		...Object.keys( currentArgs )
+	);
+	const newUrl = addQueryArgs( currentUrlWithoutArgs, {
+		...currentArgs,
+		patternId,
+	} );
+	window.history.pushState( null, '', newUrl );
```

Comment:
I think you could move all of this into the `editInPatternOnlyMode` function, that way you don't need the `editSourcePattern` function.

By doing that the editor is responsible for its own routing rather than the block.